light-client/src/light_client.rs

```diff
@@ -195,31 +194,34 @@ impl LightClient {
                 return Ok(trusted_state);
             }
 
-            // Fetch the block at the current height from our peer
-            let current_block = self.get_or_fetch_block(current_height, state)?;
+            // Fetch the block at the current height from the light store if already present,
+            // or from the primary peer otherwise.
+            let (current_block, status) = self.get_or_fetch_block(current_height, state)?;
 
-            // Validate and verify the current block
-            let verdict = self
-                .verifier
-                .verify(&current_block, &trusted_state, &options);
+            if status != Status::Trusted {
```

Comment:
If I understood correctly given that we should always have just one verifier per peer I'm not sure if we are ever going to be in a situation where an intermediate header moves to the `Trusted` state concurrently. And if say that there might be another concurrent `verifier/supervisor` updating the state of the blocks, this if does not prevent the status of this specific light block to be downgraded. I think that this check might make more sense if it was combined with a `bail` as the current concurrency model does not really allow us to get into this state.

Reply:
It's not about concurrent access, more about accidentally stumbling upon a block we already trust (has status `Trusted`) during verification/bisection and then downgrading it to `Verified`.

See the updated issue at #418 for a hopefully clearer description of the problem.